casper/src/main/scala/coop/rchain/casper/util/rholang/InterpreterUtil.scala

```diff
@@ -86,67 +61,91 @@ object InterpreterUtil {
+    val internalDeploys = ProtoUtil.deploys(block).map(InternalProcessedDeploy.fromProcessedDeploy)
+    val timestamp       = block.header.timestamp
+    val blockNumber     = block.body.state.blockNumber
+    for {
+      invalidBlocksSet <- dag.invalidBlocks
+      unseenBlocksSet  <- ProtoUtil.unseenBlockHashes(dag, block)
+      seenInvalidBlocksSet = invalidBlocksSet.filterNot(
+        block => unseenBlocksSet.contains(block.blockHash)
+      ) // TODO: Write test in which switching this to .filter makes it fail
+      invalidBlocks = seenInvalidBlocksSet
+        .map(block => (block.blockHash, block.sender))
+        .toMap
+      _         <- Span[F].mark("before-process-pre-state-hash")
+      blockData = BlockData(timestamp, blockNumber)
+      isGenesis = block.header.parentsHashList.isEmpty
+      replayResult <- runtimeManager.replayComputeState(preStateHash)(
+                       internalDeploys,
+                       blockData,
+                       invalidBlocks,
+                       isGenesis
+                     )
+      tsHash = ProtoUtil.tuplespace(block)
+      result <- handleErrors(tsHash, replayResult)
+    } yield result
+  }
+
+  private def handleErrors[F[_]: Sync: Log](
+      tsHash: ByteString,
+      result: Either[ReplayFailure, StateHash]
   ): F[BlockProcessing[Option[StateHash]]] =
-    runtimeManager
-      .replayComputeState(preStateHash)(internalDeploys, blockData, invalidBlocks, isGenesis)
-      .flatMap {
-        case Left((Some(deploy), status)) =>
-          status match {
-            case InternalErrors(exs) =>
-              BlockStatus
-                .exception(
-                  new Exception(s"Internal errors encountered while processing ${PrettyPrinter
-                    .buildString(deploy)}: ${exs.mkString("\n")}")
-                )
-                .asLeft[Option[StateHash]]
-                .pure
-            case UserErrors(errors: Seq[Throwable]) =>
-              Log[F].warn(s"Found user error(s) ${errors.map(_.getMessage).mkString("\n")}") >>
-                none[StateHash].asRight[BlockError].pure
-            case ReplayStatusMismatch(replay: DeployStatus, orig: DeployStatus) =>
-              Log[F].warn(
-                s"Found replay status mismatch; replay failure is ${replay.isFailed} and orig failure is ${orig.isFailed}"
-              ) >>
-                none[StateHash].asRight[BlockError].pure
-            case UnknownFailure =>
-              Log[F].warn(s"Found unknown failure") >>
-                none[StateHash].asRight[BlockError].pure
-            case UnusedCommEvent(_) =>
-              Sync[F].raiseError(new RuntimeException("found UnusedCommEvent"))
-          }
-        case Left((None, status)) =>
-          status match {
-            case UnusedCommEvent(_: ReplayException) =>
+    result.pure.flatMap {
+      case Left((Some(deploy), status)) =>
+        status match {
+          case InternalErrors(exs) =>
```

Comment:
To me, it looks like `InternalErrors` is just used as a catch-all for non-Interpreter errors found during the evaluation of a deployment, and that feels wrong. We should check for the errors we expect and throw `BugFoundError` when we see one that's unexpected. Curious your thoughts.

Reply:
Thanks for sharing this! I'd investigate and address that in a subsequent PR though.